### DISCUSSION TITLE
flicking light

### DISCUSSION THREAD

**Discussion by SzwedaM** [MichielVanwelsenaere/HomeAutomation.CoDeSys3#94] [Q&A]
Hello,
I have started using your e!Cockpit project and I am happy with results. I use Wago PLC (PFC200) to control my home. I have utilized the mqtt messages to control my lights from Home Assistance. It works pretty good, but I have one a problem. Sometimes when I try to switch off one of my lights using a phisical switch than the light starts flicking. A control relay is flicking until I press the switch again. It happens rarely  (ones a day or every couple days) for random lights. I use the function block FB_OUTPUT_SWITCH_MQTT to control my lights.

What am I doing wrong? What should I check?

Thank you for your assistance.

Kindly regards,
Mariusz

**Comment by MichielVanwelsenaere** [comment]
Hi Mariusz,

Is this occuring for all of your lights or just one light specifically?

What kind of relays are you using?

Ik suppose the MQTT messages aren't indicating that the light is switching on and off rapidly?

It's not really my area of expertise but I think you are experiencing an electrical problem caused by 'back emf' on the relays. Some googling on 'back emf' should ellaborate it more and provide possible solutions. Ik believe wiring a diode in should fix your problem.

Let me Know how things go. 

Best regards,
Michiel

**Reply by SzwedaM** [reply]
Hi Michiel, 

Thank you for the quick response. I will try to answer your questions in 3 points:

**Is this occuring for all of your lights or just one light specifically?**
- Yes, I can say that this issue is occurring for all of my lights. I would like to add that I use Philips lamps, Ikea bulbs and other no-name lamps for my home installation.

**What kind of relays are you using?**
- I use mechanical relays with fitted diodes for EMC suppression. They are good quality relays.  I do not think there is a problem with EMC in my hardware. This hardware setup had been working for 1 year without any issue.  It was very reliable until I decided to connect my system Wago PLC with Home Assistant using MQTT communication. 
Structure of my program was very similar, so It was easy to adopt your project. 

**Ik suppose the MQTT messages aren't indicating that the light is switching on and off rapidly?**
- Last time I noticed the same time stamps for the light switching on and off in HA history.

Best regards,
Mariusz

**Reply by MichielVanwelsenaere** [reply]
Thanks for the ellaborate answer.
So the fluctuating output can be seen in the MQTT messages as well? (Rapid on/off messages for the same output?)

If we consider the output switch logic as foolproof let's have a look on the input side. The behavior you are describing could be caused by caused by noice on the input side as well.

Could you paste the piece of ST for any of output FB that sometimes experience the behavior? Is the input digital signal wired directly to the output FB of are you using a FB to preprocess the input signal? (Like FB_INPUT_PUSHBUTTON)

**Reply by SzwedaM** [reply]
Recently I noticed flickering light in my bedroom. The piece of program code for this room is attached in PDF file below.

[light.pdf](https://github.com/MichielVanwelsenaere/HomeAutomation.CoDeSys3/files/6988343/light.pdf)

**Reply by MichielVanwelsenaere** [reply]
Good to see that you are using some debounce logic. I have two suggestions for troubleshooting further;
1. increase the debounce time in the existing logic
2. adopt to use[ FB_INPUT_PUSHBUTTON_MQTT](https://github.com/MichielVanwelsenaere/HomeAutomation.CoDeSys3/blob/master/docs/FunctionBlocks/FB_INPUT_PUSHBUTTON_MQTT.md) for your input processing to verify if the problem persists with that FB as well. (debounce logic build in)

If possible adopt one light with suggestion 1 and another one with suggestion 2. (I assume you have two lights experiencing the problem). Any particular reason you are using the Wago FB and not the FB from the project for input push button processing? Asside from any reason; let's just try it out to hopefully rule something out. The FB from the project have tried & tested by many users (of which none reported flickering lights) so that would rule out a software problem.

**Reply by SzwedaM** [reply]
Thanks for your suggestions.

I was using a higher debounce time (40 ms) for my coding before, but it didn't help it. So, I put this time back to 20 ms. Second question - I have no reason I am not using your function for input push button processing. I will try to utilize this function for testing.

I'm not home now, so I will be able to test the modified code when I return in 6 weeks. I will let you know if it solves the problem.
